feat: add co-authored users based on pull request commits - issue 425

jonathanrib3 opened the pull request:
### Problem

As previously addressed on issue #425 it would be nice to know everyone which contributed to a PR, not just the author.

### Solution

So i got into a solution that consists of getting the pull request parsing the necessary data from its html url, getting the number, repository owner and name. With these in hands through Github API Client i could access the endpoint that retrieves all commits from that specific PR. 

To get all the coauthors on these commits i considered that

1 - The co authors could be specified on the commits messages through the pattern `Co-authored-by: github name or username <github_email>`. To get these data i used a regex.
2 - Every commit object has a `committer` key which contains, among other data, the committer's login. So if  the login's different  from the pull request's author, that committer is a co author.

So i iterated over the commits fetching these data and removed every nil and duplicated value and, after that, searched for their login on database. In case the co authors are coming from commit's message, if the search by user login (which can come as his name) returns nil, then i search for his email (if that search returns nil then the user's not considered a co author).  After that we're able to know the ids from the co authors and, to get all contributors ids, which includes the own pull request author, i simply concatenated the co authors ids and the pull request author id.

With these data in hand then the service that's responsible for saving contributions, associates those contributors with the saved contribution.

**Demonstration**
_I'm going to consider [Pedro Ramos' PR#536](https://github.com/Codeminer42/Punchclock/pull/536) which is co authored by Antonio Paulino and was created two weeks ago._

_Importing contributions:_

![rake running](https://github.com/Codeminer42/Punchclock/assets/50427117/86a8f868-0c3a-4c8d-a76d-ab17897dad97)

_My contributions :_

![image](https://github.com/Codeminer42/Punchclock/assets/50427117/6899cf5c-b9cd-4c54-8b7e-22bddfbad218)

_Contributions index page on /admin:_

![image](https://github.com/Codeminer42/Punchclock/assets/50427117/59e0e508-acd4-4f26-92fd-96e788157df6)

_Contributions show page on /admin:_

![image](https://github.com/Codeminer42/Punchclock/assets/50427117/5e353e4f-f150-4df4-abd6-e424133533e9)

thiagofportella commented on app/services/github/contributions/wrappers/pull_request.rb:
```diff
@@ -16,12 +19,78 @@ def repository_name
+        end
+
+        def co_authors
+          @co_authors ||= co_authors_ids_by_messages.concat(co_authors_ids_by_committers).uniq.compact
```
This will only leave `nil` values out. If you wish to also remove any empty strings, you could do something like this:
```suggestion
          @co_authors ||= co_authors_ids_by_messages.concat(co_authors_ids_by_committers).uniq.compact.reject(&:blank?)
```
![Captura de Tela 2023-10-11 às 17 04 09](https://github.com/Codeminer42/Punchclock/assets/52546576/f6d51a5e-f423-475c-9f9e-843e768a2646)

thiagofportella commented on app/services/github/contributions/wrappers/pull_request.rb:
```diff
@@ -16,12 +19,78 @@ def repository_name
+
+            data = nil
+
+            unless !commit_message.match?(CO_AUTHOR_REGEX)
```
```suggestion
           if commit_message.match?(CO_AUTHOR_REGEX)
```

A bit more legible, IMO. What are your thoughts?

jonathanrib3 replied:
Indeed, it's actually more intuitive that way

thiagofportella commented on app/services/github/contributions/wrappers/pull_request.rb:
```diff
@@ -16,12 +19,78 @@ def repository_name
+        end
+
+        def get_co_authors_data_by_commit_messages
+          co_authors_data = pull_request_commits.map do |commit|
+            commit_message = commit.dig("commit", "message")
+
+            data = nil
+
+            unless !commit_message.match?(CO_AUTHOR_REGEX)
+              data = commit_message.scan(CO_AUTHOR_REGEX).flatten
+              data.first.strip!
+            end
+
+            data
+          end.compact
```
Perhaps you could replace `map {...}.compact` in favor of `filter_map`.
```suggestion
          co_authors_data = pull_request_commits.filter_map do |commit|
            commit_message = commit.dig("commit", "message")

            data = nil

            unless !commit_message.match?(CO_AUTHOR_REGEX)
              data = commit_message.scan(CO_AUTHOR_REGEX).flatten
              data.first.strip!
            end

            data
          end
```

https://www.rubydoc.info/stdlib/core/Enumerable:filter_map

jonathanrib3 commented:
Thanks for the tips! I adjusted those points on 1e82d13

nehamamandelbaum reviewed (commented):
Just left a small suggestion.

jonathanrib3 commented on app/models/contribution.rb:
```diff
@@ -21,6 +21,10 @@ class Contribution < ApplicationRecord
   belongs_to :reviewed_by, class_name: "User", foreign_key: "reviewer_id", optional: true
 
+  has_many :contributions_users, dependent: :destroy
+  has_many :users, through: :contributions_users
```
I thought about that but i was getting confused on how i would associate all the authors to the contribution through their id. I personally never used a HABTM association before, i should search for some examples

jonathanrib3 replied:
So searching for some examples and testing simple demos i decided to keep that way at least for now due to the amount of reworking that this would cause, plus for the feature tests on admin/contributions i would need to modify the `users` factory

nehamamandelbaum replied:
No problem! It was just a suggestion, the code is great as it is!

jonathanrib3 replied:
Thanks to your suggestion i could finally understand how HABTM experimenting on rails console, thanks for your attention!

adbatista replied:
HABTM would fit here

adbatista commented on spec/services/github/contributions/create_spec.rb:
```diff
@@ -46,6 +48,11 @@
       end
 
+      it 'associates users with contributions' do
+        expect(ContributionsUser).to receive(:create).twice
```
receiving twice, but what are the params, are they correct?

adbatista commented on spec/services/github/contributions/create_spec.rb:
```diff
@@ -46,6 +48,11 @@
 
+      it 'associates users with contributions' do
+        expect(ContributionsUser).to receive(:create).twice
+        call_create
```
call the subject directly here, it will be more explicit

jonathanrib3 replied:
In this case `call_create` is the subject, it would be better to replace by `subject` only?

adbatista commented on app/views/contributions/index.html.erb:
```diff
@@ -29,6 +30,11 @@
                   <td><%= contribution.pr_state.nil? ? '-' : contribution.pr_state_text %></td>
                   <td><%= contribution.description.present? ? t('contributions.description_filled') : '-' %></td>
+                  <td> 
+                    <% contribution.users.each do |user| %>
+                      <%= "#{ user.name }; " %>
+                    <% end %>
```
```suggestion
                    <%= contribution.users.map(&:name).join("; ") %>
```

adbatista commented on app/services/github/contributions/wrappers/pull_request.rb:
```diff
@@ -16,12 +19,78 @@ def repository_name
+
+            if co_author_id.nil? 
+              co_author = User.find_by(email: co_author_data.last)
+              co_author_id = co_author.present? ? co_author.id : nil
```
```suggestion
              co_author_id = co_author&.id
```

adbatista commented on app/services/github/contributions/wrappers/pull_request.rb:
```diff
@@ -16,12 +19,78 @@ def repository_name
+          Set.new(co_authors_data)
+        end
+        
+        def co_authors_ids_by_messages
```
Why not return the User instead of only the ID?

adbatista commented on app/services/github/contributions/wrappers/pull_request.rb:
```diff
@@ -16,12 +19,78 @@ def repository_name
+          co_authors_data = pull_request_commits.map do |commit|
+            login = commit.dig("committer", "login")
+
+            login unless pull_request.user.login == login
```
I did not get the need of this

jonathanrib3 replied:
This covers the second scenario in which a co author is not defined on commit's message. I consider every committer that's not the PR's author as a co author too.

adbatista commented on app/services/github/contributions/wrappers/pull_request.rb:
```diff
@@ -16,12 +19,78 @@ def repository_name
+        end
+
+        def get_co_authors_data_by_commit_messages
+          co_authors_data = pull_request_commits.filter_map do |commit|
+            commit_message = commit.dig("commit", "message")
+
+            data = nil
+
+            if commit_message.match?(CO_AUTHOR_REGEX)
+              data = commit_message.scan(CO_AUTHOR_REGEX).flatten
+              data.first.strip!
+            end
+
+            data
+          end
+          Set.new(co_authors_data)
```
```suggestion
          co_authors_data = pull_request_commits.filter_map do |commit|
            commit_message = commit.dig("commit", "message")

            if commit_message.match?(CO_AUTHOR_REGEX)
              data = commit_message.scan(CO_AUTHOR_REGEX).flatten
              data.first.strip!
            end
          end
          
          co_authors_data.to_set
```

adbatista commented on app/services/github/contributions/wrappers/pull_request.rb:
```diff
@@ -2,10 +2,13 @@ module Github
     module Wrappers
       class PullRequest
-        def initialize(pull_request:, engineers:, repositories:)
+        CO_AUTHOR_REGEX = %r{Co-authored-by:\s?([À-ÿ\w\s.,'-]*)\s?<(.*)>}
```
Why are you using the %r syntax for regex?

adbatista reviewed (changes requested):
Specs breaking

jonathanrib3 commented:
Morning! I fixed the specs and refactored the review points on 1cdc5f57764763154db055a3fb0a204d2c65c2d1 (refactoring contributions users association to HABTM), e66da1be573a1466def1f773b428294c5eaa2067 (refactoring pull request, create contribution service and contributions' index view according to the review), d6ef14f043c065149e8fb80c7ab9cc6a7c3f448e (fixing contributions users association where it was breaking), and 55f26b21ed9fa30f9d6ae8f6cec3e5112a493d33 (fixing breaking specs)

jonathanrib3 commented:
Updating: Added a rake task that migrates the old association between contributions and users to the new one on ace0adcbc40fde9bedfdb11e13432f6e00c6d3d2 and added a fix to new contributions page on 88103bc34e49e7db30b360ae2daedd180dc58372 - it was breaking due to the new association too.
